Feature/global notification setting

theck13 opened the pull request:
### Fix
Add a global notification preference to the ***Notification Settings*** screen under the ***Sights and Sounds*** section to allow users to enable/disable notifications on their device.  See the screenshots below for illustration.

![notification_settings](https://user-images.githubusercontent.com/3827611/30082538-688648c8-9248-11e7-9725-59582002cfc4.png)

### Test
1. Go to ***Me*** tab.
2. Tap ***Notification Settings*** option.
3. Notice ***Receive notifications*** setting.
4. Notice ***Receive notifications*** setting is on.
5. Notice all ***Sights and Sounds*** settings are enabled.
6. Like a post from another account.
7. Notice notification is received.
8. Tap ***Receive notifications*** setting.
9. Notice ***Receive notifications*** setting is off.
10. Notice all ***Sights and Sounds*** settings are disabled.
11. Like a post from another account.
12. Notice notification is not received.

akirk commented:
So far I think we recommended to use the Android setting for blocked notifications, does this add another layer of complexity in terms of support?

theck13 commented:
This is effectively the same as the Android notification setting, but the settings are all in one place (i.e. the WordPress app) rather than two places (i.e. the WordPress app and the Settings app).

With these changes, it is possible that a user can turn on notifications within the WordPress app and turn off notifications within the Settings app.  Once we update the app `minSdkVersion` to 19, we can use the [NotificationManagerCompat.areNotificationsEnabled()](https://developer.android.com/reference/android/support/v4/app/NotificationManagerCompat.html#areNotificationsEnabled()) method to check if notifications are disabled by Android and disable the notification settings within WordPress.  Unfortunately, I don't see that happening for a while.  In the meantime, it can cause another layer of complexity in terms of support.

theck13 commented:
After discussing the changes with @folletto, it was decided to update the wording of the preference from ***Receive notification*** to ***Allow app notifications*** which was done with 7ffacb7462560fa41ba439f020ce07ce1f3ed01b.  A summary stating ***Disabling this will block all app notifications, regardless of type*** was added to the preference as well with 7c7b56f170f6148c919df790375009c668d19f71.  These two labels will need copy review and are shown in the screenshot below.

![settings_wording](https://user-images.githubusercontent.com/3827611/30129870-97a9573c-9304-11e7-998c-203b74487448.png)

@mattmiklic suggested adding a link from the Settings app to the WordPress app to follow [Material guidelines for notification settings](https://material.io/guidelines/patterns/notifications.html#notifications-settings-priority).  That was added with 720fa7d333f9ffb8ccb3957bc3afdfedcce5c442.  The link is signified by the [`ic_settings`](https://material.io/icons/#ic_settings) icon on the Settings > Notifications > WordPress screen for Nougat.  See the screenshots below for illustration.

![settings_link](https://user-images.githubusercontent.com/3827611/30130471-7e3219a4-9306-11e7-9cb0-44b236876a1e.png)

The link from the Settings app to the WordPress app is signified by the ***Additional settings in the app*** button on the Settings > Apps & notifications > Notifications > Notifications > WordPress screen for Oreo.  Since the WordPress app targets Nougat, the ***Additional settings in the app*** button will not be shown on a device running Oreo.  Once we update the [`targetSdkVersion`](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/build.gradle#L41) to 26, the ***Additional settings in the app*** button will be shown.  See the screenshots below for reference.

![settings_oreo](https://user-images.githubusercontent.com/3827611/30130476-85cba34c-9306-11e7-8ab5-0313cb9e8e8f.png)

michelleweber commented:
The "Disabling this will block..." text should have a period at the end.

theck13 commented:
> The "Disabling this will block..." text should have a period at the end.

I updated the summary text in e88ac00fc1d5af50c9562ce0dd00778f1786cbf8.

folletto commented:
I think we're good design wise. :)

theck13 commented:
@michelleweber, I [added the period to the preference summary](https://github.com/wordpress-mobile/WordPress-Android/pull/6625#issuecomment-327594504).  Is the wording good to go now?

michelleweber commented:
Yup, that was my only change.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsFragment.java:
```diff
@@ -517,6 +518,15 @@ public boolean onPreferenceTreeClick(PreferenceScreen preferenceScreen, @NonNull
         }
 
+        // Track when users enable and disable global notification setting (i.e. "Allow app notifications").
+        if (preference instanceof SwitchPreference&& preference.getKey().equalsIgnoreCase(getString(R.string.wp_pref_notification_receive))) {
```
Minor nit: we need a space before `&&` here.

theck13 replied:
Good eye!  That would have bothered me so much that I would have created another pull request just to fix it.  I added the space in .Good eye!  That would have bothered me so much that I would have created another pull request just to fix it.  I added the space in 85b326cbd7ec9af2d1deed0687ad7267600e4baa.

nbradbury commented:
Code looks good, but I think this new setting is so buried that few users will benefit from it, and some of those who find it and disable it won't remember how to re-enable it (especially if they disabled it by mistake and can't figure out why notifications aren't showing).

If I wanted to enable/disable notifications in an app, I'd look for that choice at the very top of the app's notification settings and wouldn't expect to have to scroll down to a "Sights and Sounds" section to find it. I'm sure I've missed some discussion about this, but is there any reason not to make this the topmost notification setting?

theck13 commented:
That's a good point.  I don't think adding the switch as the topmost setting was discussed.  Were you thinking about something like this which would enable all preferences on the Notification Settings screen when on and disable all preferences when off?

![Topmost Switch Setting](https://storage.googleapis.com/material-design/publish/material_v_11/assets/0Bzhp5Z4wHba3blBlR3RGYUtjbjA/settings_switch2.png)

nbradbury commented:
Yes, exactly. If I wanted to completely disable notifications in an app, I'd look for that choice at the very top of the notification settings and I wouldn't care about any other notification settings.

theck13 commented:
That makes sense to me.  What do you think, @folletto?  This would mean moving the global/master switch to just below the toolbar as shown in [the screenshot above](https://camo.githubusercontent.com/c4208220b38224a1dbb2ba6539e94884df4d6e78/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6d6174657269616c2d64657369676e2f7075626c6973682f6d6174657269616c5f765f31312f6173736574732f30427a6870355a347748626133626c426c523352475955746a626a412f73657474696e67735f737769746368322e706e67) and removing the ***Allow app notifications*** switch from the ***Sights and Sounds*** section.  Would that be too subtle?  Turning the switch off will dim all the settings on the screen similar to those in the ***Sights and Sounds*** section of the [After (Disabled) screenshot](https://user-images.githubusercontent.com/3827611/30082538-688648c8-9248-11e7-9725-59582002cfc4.png).

theck13 commented:
See the mockups below for illustration.  The ***Future*** view shows a possible iteration of this design once we update the app `minSdkVersion` to 19.  That will allow us to check if notifications have been disabled by Android as mentioned [here](https://github.com/wordpress-mobile/WordPress-Android/pull/6625#issuecomment-327536277) and link the user to the Android notification settings with the ***SETTINGS*** action.

![mockup](https://user-images.githubusercontent.com/3827611/30235630-c31c33ca-94c7-11e7-9e1f-f36769d6bd99.png)

folletto commented:
It makes a lot of sense in terms of page hierarchy (IA wise it's the right placement), but it brings back even more what we discussed about it being too easy to disable. Its placement was meant to be unconspicuous so it was visible for someone looking for it, but not too visible so people would just switch it off without thinking twice.

Notifications are one of the most important reasons for the app to be used, if it seems too normal to disable it, I worry too many people will just disable and forget.

If we want it at the top, we should introduce a message when someone tries to switch off like you suggested in the chat Tyler, with a clarification message to make sure that's exactly what they mean to do: 

* "You're about to disable all notifications from the app, regardless of type.\n\nAre you sure you want to disable all notifications?" 
* ( Back | Disable All )

(pending Copy review)

theck13 commented:
> Notifications are one of the most important reasons for the app to be used, if it seems too normal to disable it, I worry too many people will just disable and forget.

I agree; notifications are important.  Part of that importance is also notifying users in the right way so that they don't want to disable all notifications from the app.  Having a global/master notification switch at the app level will allow us to track when users disable notifications and rethink our notification behavior.  If notifications are disabled at the system level, we can't track that on devices prior to API 19 and the app currently supports devices back to API 16.

> If we want it at the top, we should introduce a message when someone tries to switch off

Looking at the Settings app as an example of the topmost switch interface, toggling the global/master switch off displays a simple text view explaining the setting behavior.  What do you think about something like that, @folletto?  See the screenshots below of the Wi-Fi screen on Nougat and Oreo for example.

![toggle](https://user-images.githubusercontent.com/3827611/30243226-6f51a996-9562-11e7-8411-16f03190d676.png)

folletto commented:
> What do you think about something like that

It would be analogous to a warning message (with even more available space), so it could work.

theck13 commented:
Here is a mockup with the simple text view on the WordPress app and a screenshot of the Settings app.

![mockup](https://user-images.githubusercontent.com/3827611/30282359-4c61c12a-96d2-11e7-8dfe-a1e5f42aaace.png)

Note:
- A secondary toolbar is used to mimic the layout for other screens within the WordPress app (e.g. Reader, Stats).
- Off hides the ***Search*** action from the primary toolbar since there is nothing to search.  The action can also be disabled rather than hidden as an alternative design.
- A link to the Notifications screen within the Settings app can be added in the future similar to the ***Wi-Fi preferences*** link in the screenshot on the right once the [NotificationManagerCompat.areNotificationsEnabled()](https://developer.android.com/reference/android/support/v4/app/NotificationManagerCompat.html#areNotificationsEnabled()) method can be used to check if notifications are disabled by Android (i.e. API 19).

@folletto, can you review the design?  @michelleweber, can you review the copy?

folletto commented:
Looks good.

Given we have no more space constraints, we might add:

> You can consider turning notifications on while fine-tuning which kind of notification you receive.

folletto commented:
Looks good.

Given we have no more space constraints, we might add:

> You can consider turning notifications on while fine-tuning which kind of notification you receive.

michelleweber commented:
No changes from me.

theck13 commented:
> You can consider turning notifications on while fine-tuning which kind of notification you receive.

I added the sentence to the text view message in the screenshot below.

![mockupdate](https://user-images.githubusercontent.com/3827611/30388119-8f8f97f0-9863-11e7-9186-cc9342602e29.png)

Note:  I removed the duplicated comment also.

folletto commented:
Good for me 👍

nbradbury commented:
Looks good! :shipit:
